src/Symfony/Component/HttpClient/Internal/AmpListenerV5.php

```diff
@@ -0,0 +1,203 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpClient\Internal;
+
+use Amp\Http\Client\ApplicationInterceptor;
+use Amp\Http\Client\Connection\Connection;
+use Amp\Http\Client\Connection\Stream;
+use Amp\Http\Client\EventListener;
+use Amp\Http\Client\NetworkInterceptor;
+use Amp\Http\Client\Request;
+use Amp\Http\Client\Response;
+use Symfony\Component\HttpClient\Exception\TransportException;
+
+/**
+ * @author Nicolas Grekas <p@tchwork.com>
+ *
+ * @internal
+ */
+class AmpListenerV5 implements EventListener
+{
+    private array $info;
+
+    /**
+     * @param resource|null $handle
+     */
+    public function __construct(
+        array &$info,
+        private array $pinSha256,
+        private \Closure $onProgress,
+        private &$handle,
+    ) {
+        $info += [
+            'connect_time' => 0.0,
+            'pretransfer_time' => 0.0,
+            'starttransfer_time' => 0.0,
+            'total_time' => 0.0,
+            'namelookup_time' => 0.0,
+            'primary_ip' => '',
+            'primary_port' => 0,
+        ];
+
+        $this->info = &$info;
+    }
+
+    public function completeDnsResolution(Request $request): void
```

Comment:
This one I need to figure out how we can hook instead: this used to exist on v4 but it's not on v5 anymore.
Hooking here is useful for `NoPrivateNetworkHttpClient` so that it can filter the remote host before trying to connect to it.

Reply:
@kelunik @trowski this might be an important missing hook: reading `DnsSocketConnector`, there is no easy way currently to hook between resolve() and stream_socket_client(). Providing a custom resolver doesn't help because of the `inet_pton($host)` shortcut.

Reply:
Thinking loud: when the resolver returns multiple records, does Amp try to connect to all of them and the fastest one wins? Is that the reason for not providing this hook?

Reply:
Sorry for the spam: `Could not connect to host, try next host in the list.` so they are tried sequentially. Hook is possible :)